File: Busca_Largura_RestaUm_Serial/Java/BuscaLargura.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */


import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

/**
 *
 * @author targueriano
 */
public class BuscaLargura {
 
    
    ArrayList<Nodo> I = new ArrayList<>(); //vetor para armazenar os nodos já visitados
    ArrayList<Nodo> E = new ArrayList<>(); //os nodos filhos
    RestaUm resta;
    
    private Nodo nodoAtual;
    
    public int buscaLargura(){
        //estadoAtual = estadoInicial;
        resta = new RestaUm();
        
        E.add(new Nodo(resta.getEstadoAtual()));
        int cont = 0;
        while(!E.isEmpty()){
            System.out.println(cont++);
            nodoAtual = resta.removePrimeiro(E);
            //resta.imprimir(nodoAtual.estado);
            if(Arrays.deepEquals(nodoAtual.estado, resta.getEstadoFinal())){
                  resta.caminho(nodoAtual);
                  return 1;
            }
            
            I.add(nodoAtual); //add o nodo ja analisado
            
            resta.insereNoFinal(nodoAtual,I,E); //insere na fila 
            //ordena conforme o estado mais proximo do final
            Collections.sort(E, (Nodo o1, Nodo o2) -> (resta.valor(o1.estado) - resta.valor(o2.estado))); //heuristica
            
            
        }
        return -1;
                
    }
    
    
    public static void main(String[] args) {
        BuscaLargura bl = new BuscaLargura();
        bl.buscaLargura();
    }
}

File: Busca_Largura_RestaUm_Serial/Python/Cython/busca_largura.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Created on Fri May 20 15:36:02 2016

@author: targueriano
"""

from Nodo import Nodo
from RestaUm import RestaUm
from operator import attrgetter
#import numpy

class Busca_largura:
    I = list()
    E = list()


    def buscaLargura(self):
        resta = RestaUm()
        nodo = Nodo(resta.estadoAtual)
        self.E.append(nodo)
        cont = 0
        while(self.E):
            print(cont)
            cont += 1

            nodoAtual = resta.removePrimeiro(self.E)

            if nodoAtual.estado == resta.estadoFinal:
                resta.caminho(nodoAtual)
                return 1
            """
            if numpy.array_equal(nodoAtual.estado,resta.estadoFinal):
                resta.caminho(nodoAtual)
                return 1
            """

            self.I.append(nodoAtual)

            resta.insereNoFinal(nodoAtual, self.I, self.E)
            #verificar metodo collections
            #ordenar E conforme nodo com menor valor (resta.valor())
            self.E.sort(key=attrgetter("valor"))

        return -1




if __name__ == "__main__":
    bl = Busca_largura()
    bl.buscaLargura()



